src/go/rpk/pkg/config/params.go

```diff
@@ -502,61 +503,95 @@ func (c *Config) Write(fs afero.Fs) (rerr error) {
 	return rpkos.ReplaceFile(fs, location, b, 0o644)
 }
 
-// LocateRedpandaConfig returns the filepath location of redpanda.yaml.
-func (p *Params) LocateRedpandaConfig(fs afero.Fs) (string, error) {
-	return p.locateConfig(fs, false)
-}
-
-func (p *Params) locateConfig(fs afero.Fs, includeRpk bool) (string, error) {
-	paths := []string{p.ConfigFlag}
-	if p.ConfigFlag == "" {
-		paths = nil
-		if includeRpk {
-			if configDir, _ := os.UserConfigDir(); configDir != "" {
-				paths = append(paths, filepath.Join(configDir, "rpk", "rpk.yaml"))
-			}
-		}
-		if cd, _ := os.Getwd(); cd != "" {
-			paths = append(paths, filepath.Join(cd, "redpanda.yaml"))
-		}
-		paths = append(paths, filepath.FromSlash(DefaultPath))
+func readFile(fs afero.Fs, path string) (string, []byte, error) {
+	abs, err := filepath.Abs(path)
+	if err != nil {
+		return "", nil, err
 	}
-
-	for _, path := range paths {
-		// Ignore error: we only care whether it exists, other
-		// stat() errors are not interesting.
-		exists, _ := afero.Exists(fs, path)
-		if exists {
-			return path, nil
-		}
+	file, err := afero.ReadFile(fs, abs)
+	if err != nil {
+		return "", nil, err
 	}
-
-	return "", fmt.Errorf("%w: unable to find config in searched paths %v", afero.ErrFileNotFound, paths)
+	return abs, file, err
 }
 
+// readConfig reads both the rpk.yaml and redpanda.yaml files, filling in
+// c.fileLocation and c.rpkFileLocation as appropriate.
 func (p *Params) readConfig(fs afero.Fs, c *Config) error {
-	path, err := p.locateConfig(fs, true)
-	if err != nil {
+	if err := p.readRpkConfig(fs, c); err != nil {
 		return err
 	}
-
-	file, err := afero.ReadFile(fs, path)
-	if err != nil {
+	if err := p.readRedpandaConfig(fs, c); err != nil {
 		return err
 	}
+	return nil
+}
 
-	if err := yaml.Unmarshal(file, c); err != nil {
-		return fmt.Errorf("unable to yaml decode %s: %v", path, err)
+func (p *Params) readRpkConfig(fs afero.Fs, c *Config) error {
+	def, ok := DefaultRpkYamlPath()
+	path := def
+	if p.ConfigFlag != "" {
+		path = p.ConfigFlag
+	} else if !ok {
+		return errors.New("unable to load the user config directory -- is $HOME unset?")
```

Comment:
That's slightly different -- with the code as is, if a person deliberately unsets `$HOME` and uses `--config`, this code will work. With the proposal here, rpk will fail to load no matter what, even if the end user is using `--config`. I want to be sure we support CI systems that may not have a home directory.

re: yaml, we must've disabled that linter because somehow this is passing. I do somewhat prefer the non-shouting version of `Yaml` vs. `YAML` especially since the struct is so pervasive in the codebase now, but lmk what you think.